src/lib.rs

```diff
@@ -401,78 +397,79 @@ impl<'hedwig, T, P> PublishBatch<'hedwig, T, P> {
         Ok(self)
     }
 
-    /// Publish all the messages
+    /// Publish all the messages.
+    ///
+    /// The message order as seen by the publisher is not strictly preserved. Similarly, callback
+    /// will not preserve the order.
+    ///
+    /// Users of this function have an opportunity to decide how they want to handle the publishing
+    /// errors that occur. In particular, `PublishBatch` supports retrying failed messages. This
+    /// behaviour can be invoked by returning [`PublishCallbackAction::Retry`] from the closure.
     ///
-    /// Does not consume the builder. Will return `Ok` only if and when all of the messages from
-    /// the builder have been published successfully. In case of failure, unpublished messages will
-    /// remain enqueued in this builder for a subsequent publish call.
-    pub async fn publish(&mut self) -> Result<Vec<P::MessageId>, Error>
+    /// Do note, however, that some kinds of errors may not be transient. An eample of such an
+    /// error is attempting to publish a too large message with the
+    /// [`GooglePubSubPublisher`]. For errors like these retrying is most likely
+    /// incorrect as they would just fail again.
+    ///
+    /// Publisher-specific error types may have methods to make a decision or implementation of the
+    /// callback easier.
+    ///
+    /// [`GooglePubSubPublisher`]: publishers::GooglePubSubPublisher
+    pub async fn publish<F>(&mut self, mut result_callback: F)
     where
         P: Publisher,
+        F: FnMut(&ValidatedMessage, Result<P::MessageId, P::MessageError>) -> PublishCallbackAction,
```

Comment:
callbacks are gross :( they don't compose well with errors or asynchrony

What about changing this method to instead return a Stream, so that the caller is the one to drive publishing and retrying instead of a loop internal to this method? Maybe something like `Stream<Item=Result<(), RetryableMessage>>`, where the caller can choose to retry the message or drop it and continue with the stream

Reply:
Yeah callbacks are nasty, but I’m failing to come up with different way that would make the common way to handle errors in hedwig (i.e. retrying the publishing of most of the failed messages) the difficult one. In particular if the users had to juggle their own `Stream`s, they would need to juggle at least 2 `PublishBatch`es, and not neglect to reschedule the messages that have failed in some transient way. In fact, returning a `Stream` makes the wrong thing (dropping most of the messages) _far_ too trivial.

Maybe I’m just overthinking it though...

Reply:
Getting to use streams easily would require some more sweeping changes to the public API to make it effective in any way. I don’t oppose to making them, but I’ll submit those changes as followups to this PR.

Reply:
(Another thing that goes against streams: the users _must_ consume the entire stream for the library to be valid as it is written right now)

Reply:
There should be a reasonably easy way to re-introduce the failed messages to the same stream object, `requeue(failed_msg)` or something -- though I haven't thought too deeply about it.

I'm fine with changes in follow-ups. One case notably missing from PublishCallbackAction is the other common error handling of "stop processing and propagate up the call stack." I feel like that's another argument against the callback: the user has to encounter the error first in the callback, then respond with a _third_ variant of this enum, then the method should return a Result where the user encounters the same error a second time in order to propagate.